shared/src/ledger/ibc/vp/channel.rs

```diff
@@ -851,21 +916,74 @@ where
     }
 
     fn host_height(&self) -> Height {
-        self.host_current_height()
+        ClientReader::host_height(self)
     }
 
     fn host_timestamp(&self) -> Timestamp {
         match self.ctx.storage.get_block_header().0 {
-            Some(h) => Timestamp::from_datetime(h.time.into()),
+            Some(h) => h.time.into(),
             None => Timestamp::none(),
         }
     }
 
+    fn client_update_time(
+        &self,
+        client_id: &ClientId,
+        height: Height,
+    ) -> Ics04Result<Timestamp> {
+        let key = client_update_timestamp_key(client_id);
+        match self.ctx.read_post(&key) {
+            Ok(Some(value)) => {
+                // As ibc-go, u64 like a counter is encoded with big-endian
+                let ns: [u8; 8] = value
+                    .try_into()
+                    .map_err(|_| Ics04Error::implementation_specific())?;
+                let ns = u64::from_be_bytes(ns);
+                Timestamp::from_nanoseconds(ns)
+                    .map_err(|_| Ics04Error::implementation_specific())
+            }
+            Ok(None) => Err(Ics04Error::processed_time_not_found(
+                client_id.clone(),
+                height,
+            )),
+            Err(_) => Err(Ics04Error::implementation_specific()),
+        }
+    }
+
+    fn client_update_height(
+        &self,
+        client_id: &ClientId,
+        height: Height,
+    ) -> Ics04Result<Height> {
+        let key = client_update_height_key(client_id);
+        match self.ctx.read_post(&key) {
+            Ok(Some(value)) => Height::decode_vec(&value)
+                .map_err(|_| Ics04Error::implementation_specific()),
+            Ok(None) => Err(Ics04Error::processed_height_not_found(
+                client_id.clone(),
+                height,
+            )),
+            Err(_) => Err(Ics04Error::implementation_specific()),
+        }
+    }
+
     fn channel_counter(&self) -> Ics04Result<u64> {
         let key = channel_counter_key();
         self.read_counter(&key)
             .map_err(|_| Ics04Error::implementation_specific())
     }
+
+    fn max_expected_time_per_block(&self) -> Duration {
```

Comment:
I think when querying this value we should return the `Err` instead of the default value (in case the tx runs out of gas):
```suggestion
    fn max_expected_time_per_block(&self) -> Result<Duration> {
```

Reply:
Thank you for your suggestion! Yes, it should return Result. I'll propose and modify that in ibc-rs.